### PR TITLE
[fix] [pkg/stanza/fileconsumer] do not update fingerprint size when less data has been read

### PR DESCRIPTION
**Pull request by kasia-kujawa**
**Description:** <Describe what has changed.>

Do not update fingerprint bytes when less data has been read than it was read during initialisation of fingerprint bytes 

**Link to tracking Issue:** <Issue number if applicable>

 #22936 

**Testing:** <Describe what testing was performed and which tests were added.>

Manual tests - done
Unit test - in progress

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Comment by djaglowski** [comment]
@kkujawa-sumo, thanks for reporting this and opening the PR.

I believe this may be related to some observations made by @maokitty in #19767, #19428, and #20851. Unfortunately, we were not able to resolve the issues yet.

**Comment by kkujawa-sumo** [comment]
It looks like unit tests failed  for [commit](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/22937/commits/c5e96fbdbc0e88eba033ac8c6fb61ae71a2c3ae3) because of additional operation in `Read` function, to see similar problem in my environment I had to modify test to make rotation very very fast - so probably problems reported by @@maokitty show up.

@djaglowski I propose another approach to fix problem that I found (@22936) increasing scanner buffer  to fingerprint size to enable reading the whole fingerprint in single iteration, please see [this](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/22937/commits/6e9dff039ac13e5aaafdccd102732cdfe0ed64a7). What do you think about this approach?

File: pkg/stanza/fileconsumer/reader.go [hunk 1]
```diff
@@ -178,8 +179,8 @@ func (r *Reader) Read(dst []byte) (int, error) {
 	}
 	n, err := r.file.Read(dst)
 	appendCount := min0(n, r.fingerprintSize-int(r.Offset))
-	// return for n == 0 or r.Offset >= r.fileInput.fingerprintSize
 	if appendCount == 0 {
+		r.Debugw("Fingerprint does not need to be updated")
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/reader.go:183]: This is likely to create excessive logging, even at the debug level in my opinion. Is it really valuable?

File: pkg/stanza/fileconsumer/scanner.go [hunk 2]
```diff
@@ -51,6 +51,10 @@ func (ps *PositionalScanner) Pos() int64 {
 	return ps.pos
 }
 
+func (ps PositionalScanner) BufferSize(fingerPrintSize int) int {
+	return defaultBufSize
+}
+
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/scanner.go:57]: Is this needed?

File: pkg/stanza/fileconsumer/reader_factory.go [hunk 3]
```diff
@@ -80,6 +83,15 @@ func (b *readerBuilder) withFingerprint(fp *Fingerprint) *readerBuilder {
...
 
+func (b *readerBuilder) withBufferSize() *readerBuilder {
+	if b.readerConfig.fingerprintSize > defaultBufSize {
+		b.bufferSize = b.readerConfig.fingerprintSize
+	} else {
+		b.bufferSize = defaultBufSize
+	}
+	return b
+}
+
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/reader_factory.go:94]: Any reason not to always set the buffer size like this?

**Review by djaglowski — COMMENTED** [review]
It's still not clear to me _why_ this scenario fails. Can you describe the failure mechanism in more detail?

File: pkg/stanza/fileconsumer/scanner.go (hunk 2 above)
Comment (kkujawa-sumo) [reply pkg/stanza/fileconsumer/scanner.go:57]: yes, it isn't needed now

File: pkg/stanza/fileconsumer/reader.go (hunk 1 above)
Comment (kkujawa-sumo) [reply pkg/stanza/fileconsumer/reader.go:183]: It is not needed

**Comment by kkujawa-sumo** [comment]
@djaglowski  more details

This will describe what happens in [Read](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/34d555c72fc523cd9a907a3e7f01671d4b91e5d7/pkg/stanza/fileconsumer/reader.go#L173-L189) function now when [generated_log.txt](https://gist.github.com/kkujawa-sumo/60babc6790bf229c6380a6c645a936c8) is read 

in initialization fingerprint is read, 
 - length of fingerprint: 16450 (equal to length of the log file), 
 - in configuration `fingerprint_size: 17408`
 - scanner buffer size 16384 B = 16 kB

1. the 1s execution of [Read](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/34d555c72fc523cd9a907a3e7f01671d4b91e5d7/pkg/stanza/fileconsumer/reader.go#L173-L189)  function
  - at the beginning: len(r.Fingerprint.FirstBytes):  16450  offset:  0 read Bytes:  0
  - Reading of the file `r.file.Read(dst)`
    - all read Bytes from file:  16384  offset  0  appendCount  16384
    - new `fingerprint r.Fingerprint.FirstBytes = append(r.Fingerprint.FirstBytes[:r.Offset], dst[:appendCount]...)`
      so new size of fingerprint is 16384
 
2. the 2nd execution of [Read](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/34d555c72fc523cd9a907a3e7f01671d4b91e5d7/pkg/stanza/fileconsumer/reader.go#L173-L189) function
   - at the beginning: len(r.Fingerprint.FirstBytes):  16384  offset:  16330 read Bytes:  16384
   - Reading of the file `r.file.Read(dst)`
      - read Bytes:  16384  offset  0  appendCount  16384
      - all read Bytes from file:  16450 (the whole file was read), in this execution of `r.file.Read(dst)` 66 bytes was read
      - offset: 16330, appendCount  66
      - again new fingerprint is set
        `r.Fingerprint.FirstBytes = append(r.Fingerprint.FirstBytes[:r.Offset], dst[:appendCount]...)` so it is
        `r.Fingerprint.FirstBytes = append(r.Fingerprint.FirstBytes[:r.16330], dst[:66]...)`
        so new size of fingerprint is `16330 + 66 = 16396` and **54 bytes (16384 - 16330) are missed** from fingerprint 
        so last character of new fingerprint are different (in comparison of fingerprints there are not equal!) because `os.File.Read()` reads only unread bytes from the file 
        (below explanation what happens in Scan() function and why log file is correctly read)
             
 3. the3rd execution of Read function
    - at the beginning: len(r.Fingerprint.FirstBytes):  16396  offset:  16450 read Bytes:  16450
    - in [line](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/34d555c72fc523cd9a907a3e7f01671d4b91e5d7/pkg/stanza/fileconsumer/reader.go#L176)
      there is condition `int(r.Offset) > len(r.Fingerprint.FirstBytes)` and now offset is higher than the length of the fingerprint so Read will try to Read something without update of fingerprint, in [this line](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/34d555c72fc523cd9a907a3e7f01671d4b91e5d7/pkg/stanza/fileconsumer/reader.go#L177)
      
      
Checking what happens [Scan()])https://cs.opensource.google/go/go/+/master:src/bufio/scan.go;l=135) function from `bufio` package 
1. Read() reads 16384 bytes from file
2. [split()](https://cs.opensource.google/go/go/+/master:src/bufio/scan.go;l=146) splits 16384 bytes and until s.start=16330 s.end=16384 (s.end - s.start = 54 bytes)
3. buffer is copied and 54 bytes which left after splitting are copied into beginning of the buffer, see [this line](https://cs.opensource.google/go/go/+/master:src/bufio/scan.go;l=185)
new value of `s.end` is `54` see [this](https://cs.opensource.google/go/go/+/master:src/bufio/scan.go;l=186)
4. Read() reads file again to the buf[54:16384]
5. [split()](https://cs.opensource.google/go/go/+/master:src/bufio/scan.go;l=146) splits buf
6. loop finishes because of EOF


and I checked how os.File.Read() reads data from file using this simple program:
```
package main

import (
	"fmt"
	"log"
	"os"
)

func main() {

	file, err := os.Open("/Users/kkujawa/git/dmoloenda-contrib/opentelemetry-collector-contrib/test_file_read/test12345.txt") // For read access.
	defer file.Close()
	if err != nil {
		log.Fatal(err)
	}

	data := make([]byte, 5)
	count, err := file.Read(data)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("read %d bytes: %q\n", count, data[:count])

	data1 := make([]byte, 5)
	count1, err1 := file.Read(data1)
	if err != nil {
		log.Fatal(err1)
	}
	fmt.Printf("read %d bytes: %q\n", count1, data1[:count1])

}

```

for the file:
```text
0123456789
```
the output is
```
read 5 bytes: "01234"
read 5 bytes: "56789"
```
and same situation is in the code of fileconsumer, os.File.Read() reads only unread data and data which left after splitting are kept in the buffer

File: pkg/stanza/fileconsumer/reader_factory.go (hunk 3 above)
Comment (kkujawa-sumo) [reply pkg/stanza/fileconsumer/reader_factory.go:94]: Now default buffer size is set to 16 kb, default fingerprint size is set 1kb in `filelogreceiver` so the difference is significant, lower buffer for reading could impact speed of reading from the log file (more read operations with smaller buffer), higher fingerprint size could cause higher memory consumption - more data kept in memory to store fingerprint size.

Maybe consequences of setting always  `b.bufferSize = b.readerConfig.fingerprintSize` will not be visible but the issue which I observed exists only for cases when fingerprint size is higher than the buffer.

File: pkg/stanza/fileconsumer/reader.go [hunk 4]
```diff
@@ -32,6 +32,7 @@ type Reader struct {
 	processFunc   EmitFunc
 
 	Fingerprint    *Fingerprint
+	bufferSize     int
```
[the remark this answers is not in the document]
Comment (kkujawa-sumo) [reply pkg/stanza/fileconsumer/reader.go:35]: I'll change this

**Comment by djaglowski** [comment]
@kkujawa-sumo, thanks for detailing what happens in the scenario.

I am still having difficultly understanding the _root cause_ of this issue, but this stands out to me as a problem:

> 2. the 2nd execution of [Read](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/34d555c72fc523cd9a907a3e7f01671d4b91e5d7/pkg/stanza/fileconsumer/reader.go#L173-L189) function
> 
> * at the beginning: len(r.Fingerprint.FirstBytes): 16384 offset: 16330 read Bytes: 16384

Specifically, this seems to indicate that the offset has been incorrectly updated. If this is indeed the case, then I think the proposed change may only avoid the issue, rather than solve it.

**Comment by kkujawa-sumo** [comment]
> @kkujawa-sumo, thanks for detailing what happens in the scenario.
> 
> I am still having difficultly understanding the _root cause_ of this issue, but this stands out to me as a problem:
> 
> > 2. the 2nd execution of [Read](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/34d555c72fc523cd9a907a3e7f01671d4b91e5d7/pkg/stanza/fileconsumer/reader.go#L173-L189) function
> > 
> > 
> > * at the beginning: len(r.Fingerprint.FirstBytes): 16384 offset: 16330 read Bytes: 16384
> 
> Specifically, this seems to indicate that the offset has been incorrectly updated. If this is indeed the case, then I think the proposed change may only avoid the issue, rather than solve it.

Offset is set to the end of the line but amount of bytes read from the file is larger ( offset != read bytes), some data are in the buffer and they waits for next iteration of reading.

last characters read in first iteration of reading
```
2023-05-30 09:02:52.577515 +0200 CEST m=+0.009851256 line 137 log abcdefghijklmnopqrstuvwxyz1234567890, end of line 137
2023-05-30 09:02:52.577518 +0200 CEST m=+0.009853936 l
```
offset is set after `end of line 137` 
`2023-05-30 09:02:52.577518 +0200 CEST m=+0.009853936 l` is kept in the buffer
in next iteration of the reading following characters will be read
```
ine 138 log abcdefghijklmnopqrstuvwxyz1234567890, end of line 138
``` 

I think that offset is correctly set - at the end of fully processed line and the root cause is that the offset was treated as the indicator of read bytes form the file what is not true because some bytes are temporary kept in  the buffer  (when the reading finishes reading in the middle of the line).

**Comment by kkujawa-sumo** [comment]
> I looked into this some more and believe the error lies in the `Reader.Read` function. I haven't pinned down _exactly_ what was wrong. However, I rewrote the function to explicitly handle each possible case. This appears to pass the test you wrote. @kkujawa-sumo, do you mind trying this implementation?
> 
> ```go
> // Read from the file and update the fingerprint if necessary
> func (r *Reader) Read(dst []byte) (n int, err error) {
> 	n, err = r.file.Read(dst)
> 
> 	// CASE 0: Steady state
> 	//   Just return what was read.
> 	if len(r.Fingerprint.FirstBytes) == r.fingerprintSize {
> 		return
> 	}
> 
> 	// CASE 1: Fingerprint building
> 	//   The file is small or being read for the first time.
> 	//   It is the only condition in which we should update the fingerprint.
> 	//   Handled before edge cases in order to avoid unnecessary work.
> 	if len(r.Fingerprint.FirstBytes) < r.fingerprintSize && len(r.Fingerprint.FirstBytes) == int(r.Offset) {
> 		appendCount := r.fingerprintSize - len(r.Fingerprint.FirstBytes)
> 		if appendCount > n {
> 			appendCount = n
> 		}
> 		r.Fingerprint.FirstBytes = append(r.Fingerprint.FirstBytes, dst[:appendCount]...)
> 		return
> 	}
> 
> 	// The remaining cases can happen when the following has occured:
> 	//   1. The component was started with a storage extension and saved a (partial or full) fingerprint for the file.
> 	//   2. The component was restarted with the same storage extension and a different setting for 'fingerprint_size'.
> 
> 	// CASE 2: Oversized fingerprint
> 	//   The component was restarted with a decreased 'fingerprint_size'. Just keep reading.
> 	if len(r.Fingerprint.FirstBytes) > r.fingerprintSize {
> 		return
> 	}
> 
> 	// CASE 3: Undersized fingerprint
> 	//   At this point we know that the fingerprint is undersized (didn't match CASE 0 or CASE 2).
> 	//   However, we also know that the fingerprint and offset are not aligned (didn't match second condition of CASE 1).
> 	//   Therefore, we cannot update the fingerprint because we are not reading from the correct staring position.
> 	return
> }
> ```
> 
> If you agree it works, I'll continue on this tomorrow by writing a set of tests to validate each case.



@djaglowski  I'm ok with fixing the problem in `Read()` function and your proposition looks good :) 
I tested this proposition with two log files for which I saw the issue before and I don't see issue.

I also wrote test with line longer than default buffer and larger fingerprintsize - it passes:

```go
func TestReadingWithLargeFingerPrintSizeAndLineLargerThanScannerBuf(t *testing.T) {
	t.Parallel()
	tempDir := t.TempDir()

	lineLength := defaultBufSize + 1024
	fingerPrintSize := defaultBufSize + 2*1024

	// Generate log lines
	body := "0abcdefghijklmnopqrstuvwxyz"
	firstline := ""
	for len(firstline) < lineLength {
		firstline += body
	}
	firstline = fmt.Sprintf("%s\n", firstline[:lineLength-1])
	require.Equal(t, lineLength, len(firstline))

	// Create a new file
	temp := openTemp(t, tempDir)
	writeString(t, temp, firstline)

	// Create reader
	emitChan := make(chan *emitParams, 1000)
	splitterConfig := helper.NewSplitterConfig()
	f := &readerFactory{
		SugaredLogger: testutil.Logger(t),
		readerConfig: &readerConfig{
			fingerprintSize: fingerPrintSize,
			maxLogSize:      defaultMaxLogSize,
			emit:            testEmitFunc(emitChan),
		},
		fromBeginning:   true,
		splitterFactory: newMultilineSplitterFactory(splitterConfig),
		encodingConfig:  splitterConfig.EncodingConfig,
	}

	r, err := f.newReaderBuilder().withFile(temp).withBufferSize().build()
	require.NoError(t, err)

	initialFingerPrintSize := len(r.Fingerprint.FirstBytes)
	initialFingerPrint := string(r.Fingerprint.FirstBytes)
	r.ReadToEnd(context.Background())
	require.Equal(t, initialFingerPrintSize, len(r.Fingerprint.FirstBytes))
	require.Equal(t, initialFingerPrint, string(r.Fingerprint.FirstBytes))

	require.Equal(t, []byte(firstline[:lineLength-1]), readToken(t, emitChan))
}
```

**Comment by djaglowski** [comment]
Thanks for confirming @kkujawa-sumo. I've opened https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23183. Perhaps if it's merged you'll rebase to add your test?

**Comment by kkujawa-sumo** [comment]
> Thanks for confirming @kkujawa-sumo. I've opened #23183. Perhaps if it's merged you'll rebase to add your test?

@djaglowski Thank for help with this issue ❤️  
I'll check test cases which you added to the #23183 and if I see that my tests adds something new I'll open pull request ;)

**Pull request closed without merging**